client/rpc-spec-v2/src/chain_head/subscription/inner.rs

```diff
@@ -107,6 +108,75 @@ impl BlockStateMachine {
 	}
 }
 
+/// Limit the number of ongoing operations across methods.
+#[derive(Clone)]
+struct LimitOperations {
+	/// Maximum number of operations that can be executed at the same time.
+	max_operations: usize,
+	/// The number of ongoing operations for this subscription.
+	ongoing: Arc<Mutex<usize>>,
+}
+
+impl LimitOperations {
+	/// Constructs a new [`LimitOperations`].
+	fn new(max_operations: usize) -> Self {
+		LimitOperations { max_operations, ongoing: Default::default() }
+	}
+
+	/// Reserves capacity to execute at least one operation and at most the requested items.
+	///
+	/// Dropping [`PermitOperations`] without executing an operation will release
+	/// the reserved capacity.
+	///
+	/// Returns nothing if there's no space available, else returns a permit
+	/// that guarantees that at least one operation can be executed.
+	fn reserve_at_most(&self, to_reserve: usize) -> Option<PermitOperations> {
+		let mut ongoing = self.ongoing.lock();
+
+		// The maximum capacity.
+		let capacity = self.max_operations.saturating_sub(*ongoing);
+		// The number of operations we can reserve.
+		let reserved = capacity.min(to_reserve);
+
+		if reserved == 0 {
+			None
+		} else {
+			// Update the ongoing operations.
+			*ongoing = ongoing.saturating_add(reserved);
+			Some(PermitOperations { num_ops: reserved, ongoing: self.ongoing.clone() })
+		}
+	}
+}
+
+/// Permits a number of operations to be executed.
+///
+/// [`PermitOperations`] are returned by [`LimitOperations::reserve()`] and are used
+/// to guarantee the RPC server can execute the number of operations.
+///
+/// The number of reserved items are given back to the [`LimitOperations`] on drop.
+struct PermitOperations {
+	/// The number of operations permitted (reserved).
+	num_ops: usize,
+	/// The number of ongoing operations for this subscription.
+	ongoing: Arc<Mutex<usize>>,
```

Comment:
Also would be good to understand why `Mutex` is used here instead of `Arc<AtomicUsize>` or something similar?

Reply:
The `Arc<Mutex<usize>>` was shared between the frontend (permit) and the backend. On drop the values are incremented marking the release of the acquired capacity.

However, I've used `tokio::sync::Semaphore` which simplified the code a bit, and this should no longer be needed.